mindmeld/augmentation.py

```diff
@@ -0,0 +1,285 @@
+    """Paraphraser class for generating paraphrases based on language code of the app
+    (for languages other than English).
+    """
+
+    def __init__(self, app_path, config=None):
+        """Initializes a multi-lingual paraphraser.
+
+        Args:
+            app_path (str): The location of the MindMeld app
+            config (dict, optional): A config object to use. This will
+                override the config specified by the app's config.py file.
+        """
+        super().__init__(app_path=app_path, config=config)
+
+        self.torch_device = "cuda" if torch.cuda.is_available() else "cpu"
+        self.language, _ = get_language_config(app_path=app_path)
+
+        en_model_name = "Helsinki-NLP/opus-mt-ROMANCE-en"
+        self.en_tokenizer = MarianTokenizer.from_pretrained(en_model_name)
+        self.en_model = MarianMTModel.from_pretrained(en_model_name)
+        self.en_model.to(self.torch_device)
+
+        target_model_name = "Helsinki-NLP/opus-mt-en-ROMANCE"
+        self.target_tokenizer = MarianTokenizer.from_pretrained(target_model_name)
+        self.target_model = MarianMTModel.from_pretrained(target_model_name).to(
+            self.torch_device
+        )
+
+    def _translate(self, **kwargs):
+        """The core translation step for forward and back translation.
+
+        Args:
+            template (lambda func): Structure input text to model.
+            queries (list(str)): List of input queries.
+            model: Machine translation model (en-ROMANCE or ROMANCE-en).
+            tokenizer: Language tokenizer for input query text.
+            num_return_sequences (int): Number of translations generated.
+        """
+        template = kwargs.pop("template")
+        queries = kwargs.pop("queries")
+        model = kwargs.pop("model")
+        tokenizer = kwargs.pop("tokenizer")
+
+        queries = [template(query) for query in queries]
+        encoded = tokenizer.prepare_seq2seq_batch(queries, return_tensors="pt")
+        for key in encoded:
+            encoded[key] = encoded[key].to(self.torch_device)
+
+        translated = model.generate(**encoded, **kwargs)
+        translated_queries = tokenizer.batch_decode(
+            translated, skip_special_tokens=True
+        )
+        return translated_queries
+
+    def augment_queries(self, queries):
+        translated_queries = self._translate(
+            template=lambda text: f"{text}",
+            queries=queries,
+            model=self.en_model,
+            tokenizer=self.en_tokenizer,
+            num_beams=5,
+            num_return_sequences=5,
+            top_k=0,
+            temperature=1.0,
+        )
+        translated_queries = list(set(translated_queries))
+
+        back_translated_queries = self._translate(
+            template=lambda text: f">>{self.language}<< {text}",
+            queries=translated_queries,
+            model=self.target_model,
+            tokenizer=self.target_tokenizer,
+            num_beams=5,
```

Comment:
can reduce num_beams and num_sequences to 3 since we wan to generate at most 10 queries